openstack/datadog_checks/openstack/openstack.py

```diff
@@ -945,9 +919,37 @@ def get_all_servers(self, i_key, collect_all_tenants, filter_by_host=None):
                 except KeyError as e:
                     self.log.debug("Server: %s has already been removed from the cache", new_server['server_id'])
 
-        return self.server_details_by_id
+    def filter_excluded_servers(self):
+        proj_list = set([])
+        servers_copy = copy.deepcopy(self.server_details_by_id)
+        if self.exclude_server_id_rules:
+            # Filter out excluded servers
+            for exclude_id_rule in self.exclude_server_id_rules:
+                for server_id in self.server_details_by_id.keys():
+                    if re.match(exclude_id_rule, server_id):
+                        del self.server_details_by_id[server_id]
+
+        for server in self.server_details_by_id.iteritems():
+            proj_list.add(server[1].get('project_name'))
+
+        projects_filtered = pattern_filter(proj_list,
+                                           whitelist=self.include_project_name_rules,
+                                           blacklist=self.exclude_project_name_rules
+                                           )
+        for server in servers_copy.iteritems():
+            if server[1].get('project_name') not in projects_filtered:
+                # This may have already been deleted if the server belongs to both
+                # the project blacklist AND the excluded server id list
+                try:
```

Comment:
The issue here was that there are two methods to allow users to delete servers. One is the `excluded_id` thing, and another is by project name.

If a user's env has the server `{'id':1, 'project': 'blacklist_proj'}`, and they have exclude_id:1, this server will be removed from the list, then if they also have `blacklist_project: 'blacklist_*'` it will try to be deleted again. I'm avoiding deepcopying the list twice in this case by wrapping it in a try/catch.

Reply:
Yeah it seems hack-ish to me to use a try/catch here when we could avoid it with another algorithm.